rpc/providers/proxy.ts

```diff
@@ -3,44 +3,38 @@ import { deadline, deferred } from "../../deps/std/async.ts";
 import { ErrorCtor } from "../../util/mod.ts";
 import * as B from "../Base.ts";
 import { ClientHooks, ParseRawIngressMessageError } from "../common.ts";
+import { WebSocketProxy } from "./web-socket-proxy.ts";
 
 export type ProxyClientHooks<Config_ extends Config<string>> = ClientHooks<Config_, Event>;
 
-export function proxyClient<Config_ extends Config<string>>(
+export async function proxyClient<Config_ extends Config<string>>(
   config: Config_,
   hooks?: ProxyClientHooks<Config_>,
 ): Promise<ProxyClient<Config_> | FailedToOpenConnectionError> {
-  const ws = new WebSocket(config.discoveryValue);
-  const client = new ProxyClient(ws, hooks);
-  ws.addEventListener("error", client.onError);
-  ws.addEventListener("message", client.onMessage);
-  const pending = deferred<ProxyClient<Config_> | FailedToOpenConnectionError>();
-  if (ws.readyState === WebSocket.CONNECTING) {
-    const onOpenError = (e: any) => {
-      console.log({ log: e });
-      clearListeners();
-      pending.resolve(new FailedToOpenConnectionError());
-    };
-    const onOpen = () => {
-      clearListeners();
-      pending.resolve(client);
-    };
-    const clearListeners = () => {
-      ws.removeEventListener("error", onOpenError);
-      ws.removeEventListener("open", onOpen);
-    };
-    ws.addEventListener("error", onOpenError);
-    ws.addEventListener("open", onOpen);
-  } else {
-    pending.resolve(client);
+  const wsIsOpen = deferred();
+  const ws = new WebSocketProxy({
+    webSocketFactory: () => new WebSocket(config.discoveryValue),
+  });
+
+  ws.onopen = () => wsIsOpen.resolve();
```

Comment:
I wanted to keep the API very slim but I guess it would be more friendly with add/remove listeners.

Initially, I thought that this should re-dispatch the inner WebSocket events and clone those events as follows
```
const et = new EventTarget();

const ws = new WebSocket("wss://rpc.polkadot.io");

ws.addEventListener("open", (e) => {
  // this throws a DOMException InvalidStateError
  et.dispatchEvent(e);
  // cloning the event doesn't throw
  et.dispatchEvent(new e.constructor(e.type, e));
});

et.addEventListener("open", (e: Event) => {
  console.log(e);
});

```

However, this container could have its own events that don't need to match exactly to the WebSocket events.

I'll change this container to extend from `EventTarget`

Reply:
I may be missing something––I don't see why we need an `EventTarget`. Can you please elaborate from a slightly higher level?

Reply:
It would be to reuse the add/remove event listener functionality.

I noticed that the `B.Client` implements its own event listening
https://github.com/paritytech/capi/blob/11106d7b46159657e5dbf3765688ddbce360cbad/rpc/Base.ts#L54-L60

Perhaps, there is a reason to not extend from `EventTarget` or `std/node/events.ts` `EventEmitter`.
Thoughts?